File: commonutils/src/main/java/net/maxsmr/commonutils/logger/SimpleSystemLogger.java
package net.maxsmr.commonutils.logger;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import static net.maxsmr.commonutils.data.conversion.format.DateFormatUtilsKt.formatDateNoThrow;

public class SimpleSystemLogger extends BaseTagLogger {

    protected SimpleSystemLogger(@Nullable String tag) {
        super(tag);
    }

    @Override
    public void v(String message) {
        if (isLoggingEnabled()) {
            log(Level.VERBOSE, message);
        }
    }

    @Override
    public void v(Throwable exception) {
        if (exception != null) {
            v(exception.toString());
        }
    }

    @Override
    public void v(String message, Throwable exception) {
        v(message);
        v(exception);
    }

    @Override
    public void d(String message) {
        if (isLoggingEnabled()) {
            log(Level.DEBUG, message);
        }
    }

    @Override
    public void d(Throwable exception) {
        d(exception.toString());
    }

    @Override
    public void d(String message, Throwable exception) {
        d(message);
        d(exception);

    }

    @Override
    public void i(String message) {
        if (isLoggingEnabled()) {
            log(Level.INFO, message);
        }
    }

    @Override
    public void i(Throwable exception) {
        i(exception.toString());
    }

    @Override
    public void i(String message, Throwable exception) {
        i(message);
        i(exception);
    }

    @Override
    public void w(String message) {
        if (isLoggingEnabled()) {
            log(Level.WARN, message);
        }
    }

    @Override
    public void w(Throwable exception) {
        w(exception.toString());
    }

    @Override
    public void w(String message, Throwable exception) {
        w(message);
        w(exception);
    }

    @Override
    public void e(String message) {
        if (isLoggingEnabled()) {
            log(Level.ERROR, message);
        }
    }

    @Override
    public void e(Throwable exception) {
        e(exception.toString());
    }

    @Override
    public void e(String message, Throwable exception) {
        e(message);
        e(exception);
    }

    @Override
    public void wtf(String message) {
        if (isLoggingEnabled()) {
            log(Level.WTF, message);
        }
    }

    @Override
    public void wtf(Throwable exception) {
        if (exception != null) {
            wtf(exception.toString());
        }
    }

    @Override
    public void wtf(String message, Throwable exception) {
        wtf(message);
        wtf(exception);
    }

    private void log(@NotNull Level level, @Nullable String message) {
        if (message != null) {
            LogEntry logEntry = new LogEntry(level, tag, message, System.currentTimeMillis());
            if (level == Level.ERROR || level == Level.WTF) {
                System.err.println(logEntry.toString());
            } else {
                System.out.println(logEntry.toString());
            }
        }
    }

    private static class LogEntry {

        private final SimpleDateFormat SDF = new SimpleDateFormat("dd.MM.yyyy HH:mm:ss", Locale.getDefault());

        @NotNull
        private final Level level;

        @NotNull
        private final String tag;

        @NotNull
        private final String message;

        private final long timestamp;

        private LogEntry(@NotNull Level level, @NotNull String tag, @NotNull String message, long timestamp) {
            this.level = level;
            this.tag = tag;
            this.message = message;
            this.timestamp = timestamp;
        }

        @Override
        @NotNull
        public String toString() {
            return "[" + formatDateNoThrow(new Date(timestamp), SDF, null) + "] " + level.name() + " " + tag + ": " + message;
        }
    }
}

File: commonutils-jre/src/main/kotlin/net/maxsmr/commonutils/logger/SimpleSystemLogger.kt
package net.maxsmr.commonutils.logger

import net.maxsmr.commonutils.format.formatDate
import net.maxsmr.commonutils.logger.BaseLogger.Level.*
import java.util.*

class SimpleSystemLogger(tag: String) : BaseLogger(tag) {

    override fun _v(message: String) {
        log(VERBOSE, message)
    }

    override fun _v(exception: Throwable) {
        v(exception.toString())
    }

    override fun _v(message: String, exception: Throwable?) {
        v(message)
        exception?.let {
            v(it)
        }
    }

    override fun _d(message: String) {
        log(DEBUG, message)
    }

    override fun _d(exception: Throwable) {
        d(exception.toString())
    }

    override fun _d(message: String, exception: Throwable?) {
        d(message)
        exception?.let {
            d(it)
        }
    }

    override fun _i(message: String) {
        log(INFO, message)
    }

    override fun _i(exception: Throwable) {
        i(exception.toString())
    }

    override fun _i(message: String, exception: Throwable?) {
        i(message)
        exception?.let {
            i(it)
        }
    }

    override fun _w(message: String) {
        log(WARN, message)
    }

    override fun _w(exception: Throwable) {
        w(exception.toString())
    }

    override fun _w(message: String, exception: Throwable?) {
        w(message)
        exception?.let {
            w(it)
        }
    }

    override fun _e(message: String) {
        log(ERROR, message)
    }

    override fun _e(exception: Throwable) {
        e(exception.toString())
    }

    override fun _e(message: String, exception: Throwable?) {
        e(message)
        exception?.let {
            e(it)
        }
    }

    override fun _wtf(message: String) {
        log(WTF, message)
    }

    override fun _wtf(exception: Throwable) {
        wtf(exception.toString())
    }

    override fun _wtf(message: String, exception: Throwable?) {
        wtf(message)
        exception?.let {
            wtf(it)
        }
    }

    private fun log(level: Level, message: String) {
        val logEntry = LogEntry(level, tag, message, System.currentTimeMillis())
        if (level == ERROR || level == WTF) {
            System.err.println(logEntry.toString())
        } else {
            println(logEntry.toString())
        }
    }

    private class LogEntry constructor(
        private val level: Level,
        private val tag: String,
        private val message: String,
        private val timestamp: Long
    ) {

        override fun toString(): String {
            return "[" + formatDate(Date(timestamp), "dd.MM.yyyy HH:mm:ss") + "] " + level.name + " " + tag + ": " + message
        }
    }
}
